ktor-io/common/src/io/ktor/utils/io/ByteReadChannelOperations.kt

```diff
@@ -490,3 +491,173 @@ public fun ByteWriteChannel.rethrowCloseCauseIfNeeded() {
 public fun ByteChannel.rethrowCloseCauseIfNeeded() {
     closedCause?.let { throw it }
 }
+
+/**
+ * Reads bytes from the ByteReadChannel until a specified sequence of bytes is encountered or the specified limit is reached.
+ *
+ * This uses the KMP algorithm for finding the string match using a partial match table.
+ *
+ * @see [Knuth–Morris–Pratt algorithm](https://en.wikipedia.org/wiki/Knuth%E2%80%93Morris%E2%80%93Pratt_algorithm)
+ * @param matchString The sequence of bytes to look for.
+ * @param writeChannel The channel to write the read bytes to.
+ * @param limit The maximum number of bytes to read before throwing an exception.
+ * @param ignoreMissing Whether to ignore the missing byteString and return the count of read bytes upon reaching the end of input.
+ * @return The number of bytes read, not including the search string.
+ * @throws IOException If the limit is exceeded or the byteString is not found and ignoreMissing is false.
+ */
+public suspend fun ByteReadChannel.readUntil(
+    matchString: ByteString,
+    writeChannel: ByteWriteChannel,
+    limit: Long = Long.MAX_VALUE,
+    ignoreMissing: Boolean = false,
+): Long {
```

Comment:
This function will read directly to the output channel when looking for the matching bytestring, so we can rely on the coroutine suspension to keep the memory overhead in line.  I figured it would be a useful addition to the utility functions here, along with some skip functions, for when parsing directly from the bytechannel stream.